scene/main/node.cpp

```diff
@@ -2548,6 +2548,24 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 				}
 
 			} else {
+				// If property points to a node which is owned by a node we are duplicating, update its path.
+				if (value.get_type() == Variant::OBJECT) {
+					Node *property_node = Object::cast_to<Node>(value);
+					if (property_node && is_ancestor_of(property_node)) {
+						value = current_node->get_node_or_null(get_path_to(property_node));
+					}
+				} else if (value.get_type() == Variant::ARRAY) {
+					Array arr = value;
+					if (arr.get_typed_builtin() == Variant::OBJECT) {
+						for (int i = 0; i < arr.size(); i++) {
+							Node *property_node = Object::cast_to<Node>(arr[i]);
+							if (property_node && is_ancestor_of(property_node)) {
+								arr[i] = current_node->get_node_or_null(get_path_to(property_node));
+							}
+						}
+						value = arr;
+					}
+				}
```

Comment:
Wouldn't this also need to recurse into deeper array levels, dictionaries, etc.? I think we have some code around to traverse a `Variant`. If not already there, I'd suggest adding something like `PropertyUtils::visit_objects(p_variant, p_visitor_callback)`.

Reply:
As an example do you mean an array of node3Ds which is inside an array type of situation? As far as I understood, you don't have to do anything with dictionaries since they don't provide a user a way to give an Object type. So we don't have to do anything special with them until typed version of dictionary is introduced.

Reply:
@RandomShaper

Reply:
Oh, that's very true.